File: backend/services/technical_metadata.py
import re
import uuid
import datetime
from typing import Dict, Any, List, Optional
from bs4 import BeautifulSoup

class TechnicalMetadata:
    """
    Extracts and stores technical metadata for web pages:
    - Response time
    - Content length
    - HTML size
    - Word count
    - Language
    - Encoding
    - Content-Type
    - Compression
    - Cache headers
    """
    store: Dict[str, Dict[str, Any]] = {}

    @classmethod
    def extract_metadata(
        cls,
        url: str,
        html: str,
        headers: Optional[Dict[str, str]] = None,
        response_time: int = 0
    ) -> Dict[str, Any]:
        headers = {k.lower(): v for k, v in (headers or {}).items()}
        soup = BeautifulSoup(html or "", "html.parser")

        # 1. HTML size
        html_bytes = (html or "").encode("utf-8")
        html_size = len(html_bytes)

        # 2. Content Length
        content_length = html_size
        if "content-length" in headers:
            try:
                parsed_len = int(headers["content-length"])
                if parsed_len > 0:
                    content_length = parsed_len
            except ValueError:
                pass

        # 3. Word count
        for tag in soup(["script", "style", "noscript", "svg", "iframe"]):
            tag.decompose()
        body_text = soup.get_text() or ""
        words = [w for w in re.split(r"\s+", body_text.strip()) if w]
        word_count = len(words)

        # 4. Language
        html_tag = soup.find("html")
        language = html_tag.get("lang").strip() if html_tag and html_tag.get("lang") else ""
        if not language and "content-language" in headers:
            language = headers["content-language"].strip()
        if not language:
            language = "unknown"

        # 5. Encoding
        encoding = ""
        meta_charset = soup.find("meta", attrs={"charset": True})
        if meta_charset and meta_charset.get("charset"):
            encoding = meta_charset["charset"].strip()

        if not encoding:
            meta_http = soup.find("meta", attrs={"http-equiv": re.compile(r"content-type", re.I)})
            if meta_http and meta_http.get("content"):
                match = re.search(r"charset=([^\s;]+)", meta_http["content"], re.I)
                if match:
                    encoding = match.group(1)

        if not encoding and "content-type" in headers:
            match = re.search(r"charset=([^\s;]+)", headers["content-type"], re.I)
            if match:
                encoding = match.group(1)

        if not encoding:
            encoding = "utf-8"

        # 6. Content-Type
        content_type = headers.get("content-type", "text/html")

        # 7. Compression
        compression = headers.get("content-encoding", "none")

        # 8. Cache headers
        cache_headers = {
            "cacheControl": headers.get("cache-control"),
            "expires": headers.get("expires"),
            "etag": headers.get("etag"),
            "lastModified": headers.get("last-modified"),
            "pragma": headers.get("pragma"),
            "age": headers.get("age"),
            "vary": headers.get("vary")
        }

        item_id = f"tech_{str(uuid.uuid4())[:8]}"
        metadata = {
            "id": item_id,
            "url": url,
            "responseTime": response_time,
            "contentLength": content_length,
            "htmlSize": html_size,
            "wordCount": word_count,
            "language": language,
            "encoding": encoding,
            "contentType": content_type,
            "compression": compression,
            "cacheHeaders": cache_headers,
            "rawHeaders": headers,
            "createdAt": datetime.datetime.utcnow().isoformat() + "Z"
        }

        return metadata

    @classmethod
    def store_metadata(cls, metadata: Dict[str, Any]) -> Dict[str, Any]:
        cls.store[metadata["id"]] = metadata
        return metadata

    @classmethod
    def get_metadata(cls, item_id: str) -> Optional[Dict[str, Any]]:
        return cls.store.get(item_id)

    @classmethod
    def list_metadata(cls) -> List[Dict[str, Any]]:
        return list(cls.store.values())

File: src/services/technicalMetadata.ts
import * as cheerio from 'cheerio';
import { FetchResult } from './httpFetcher';

export interface CacheHeaders {
  cacheControl?: string;
  expires?: string;
  etag?: string;
  lastModified?: string;
  pragma?: string;
  age?: string;
  vary?: string;
}

export interface TechnicalMetadata {
  id: string;
  url: string;
  responseTime: number; // in ms
  contentLength: number; // in bytes
  htmlSize: number; // in bytes
  wordCount: number;
  language: string;
  encoding: string;
  contentType: string;
  compression: string;
  cacheHeaders: CacheHeaders;
  rawHeaders: Record<string, string>;
  createdAt: string;
}

export class TechnicalMetadataManager {
  private static store: Map<string, TechnicalMetadata> = new Map();

  public static extractMetadata(
    url: string,
    html: string,
    headers: Record<string, string> = {},
    responseTime: number = 0
  ): TechnicalMetadata {
    const $ = cheerio.load(html || '');

    // 1. HTML size in bytes
    const htmlSize = Buffer.byteLength(html || '', 'utf-8');

    // 2. Content Length
    let contentLength = htmlSize;
    if (headers['content-length']) {
      const parsed = parseInt(headers['content-length'], 10);
      if (!isNaN(parsed) && parsed > 0) {
        contentLength = parsed;
      }
    }

    // 3. Word Count (Extract text from body, clean scripts/styles)
    $('script, style, noscript, svg, iframe').remove();
    const bodyText = $('body').text() || $.text() || '';
    const words = bodyText.trim().split(/\s+/).filter(w => w.length > 0);
    const wordCount = words.length;

    // 4. Language
    let language = $('html').attr('lang')?.trim() || '';
    if (!language && headers['content-language']) {
      language = headers['content-language'].trim();
    }
    if (!language) {
      language = 'unknown';
    }

    // 5. Encoding / Charset
    let encoding = $('meta[charset]').attr('charset')?.trim() || '';
    if (!encoding) {
      const httpEquiv = $('meta[http-equiv="Content-Type" i]').attr('content') || '';
      const match = httpEquiv.match(/charset=([^\s;]+)/i);
      if (match) {
        encoding = match[1];
      }
    }
    if (!encoding && headers['content-type']) {
      const match = headers['content-type'].match(/charset=([^\s;]+)/i);
      if (match) {
        encoding = match[1];
      }
    }
    if (!encoding) {
      encoding = 'utf-8';
    }

    // 6. Content-Type
    const contentType = headers['content-type'] || 'text/html';

    // 7. Compression
    const compression = headers['content-encoding'] || 'none';

    // 8. Cache Headers
    const cacheHeaders: CacheHeaders = {
      cacheControl: headers['cache-control'],
      expires: headers['expires'],
      etag: headers['etag'],
      lastModified: headers['last-modified'],
      pragma: headers['pragma'],
      age: headers['age'],
      vary: headers['vary']
    };

    const id = `tech_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`;
    const metadata: TechnicalMetadata = {
      id,
      url,
      responseTime,
      contentLength,
      htmlSize,
      wordCount,
      language,
      encoding,
      contentType,
      compression,
      cacheHeaders,
      rawHeaders: headers,
      createdAt: new Date().toISOString()
    };

    return metadata;
  }

  public static storeMetadata(metadata: TechnicalMetadata): TechnicalMetadata {
    this.store.set(metadata.id, metadata);
    return metadata;
  }

  public static getMetadata(id: string): TechnicalMetadata | undefined {
    return this.store.get(id);
  }

  public static listMetadata(): TechnicalMetadata[] {
    return Array.from(this.store.values());
  }

  public static clearStore(): void {
    this.store.clear();
  }
}
